### PR TITLE
Rough Changes

### PR DESCRIPTION
**Pull request by david-vanderson**
First pass, probably far from being merge-able, but it's in the right direction I think.

1. Changes the drawing model so each layer is a separate glDrawArrayInstanced call
  - This fixes some weird alpha blending stuff that was hurting text especially
  - Also removes the layer cap

2. Changes the drawing model from hard pixel edges to smoothed
  - Needs to be made optional
  - Scaled smoothed sprites have weird gray edges, I'm looking into premultiplied alpha to fix

3. Small change to text rendering that should make the text render at the same pixel size as the screen
  - I'm not sure what this code was trying to do, but with this change I think text looks better but maybe it's less saturated?

### CODE REVIEW

**Comment by david-vanderson** [comment]
Also I noticed that if I make a 1x1 sprite it's not visible if it has integer positions.  I added some commented out code in ngl.vertext.glsl about it.  I think it's because the quad gets positioned exactly between pixels, so no fragment shader is run, but I'm not sure.  Truncating the pixel position also produces some weird effects.  Not sure what to do about this.

**Comment by jeapostrophe** [comment]
Hi David, Thanks for pushing me on this. I just pushed a version that incorporates your text change and reimplements the layer system to have no limits. I noticed in your version you tried to only upload some of them, but this is really inefficient because the most expensive activity is memory transfers, not evaluating a conditional. So in my version, I upload all the sprites and then just update the uniform for each draw call and move the loop out one layer. You'll have to rebase and reapply your other changes. Sorry for the pain.

**Comment by david-vanderson** [comment]
Thanks very much for fixing/merging it!  I had no idea about uploading only parts of them, I'm glad you fixed it.

It's no pain at all.  I'll rebase, investigate the premultiplied alpha stuff a bit more, and send another pull request towards making pixel smoothing optional.  Thanks!

**Comment by david-vanderson** [comment]
The "add more texture constants" is needed to get more than 11 layers.  With the new constants I can get up to 15 layers.  Any more gives me this:

GL context changed
You are using OpenGL '(4 1) with gl-backend-version of 3.3
max layers 80
Log of shader(4):
	ERROR: Implementation limit of 16 active fragment shader samplers (e.g., maximum number of supported image units) exceeded, fragment shader uses 17 samplers

**Comment by david-vanderson** [comment]
The rest of the commits change to smoothing and premultiplied alpha, which need to be made optional.  But I think the needed changes are all there.  I'm sure there are better ways to code it, but this seems to work visually.

File: mode-lambda/backend/gl/combine.fragment.glsl [hunk 1]
```diff
@@ -56,9 +56,7 @@ void main() {
     if ((! (pz <= 0.0))
         && (0.0 <= ey && ey <= height)
         && (0.0 <= ex && ex <= width)) {
-      fin_Color.rgb =
-          fin_Color.rgb * (1.0 - lay_Color.a)
-        + lay_Color.rgb * lay_Color.a;
+      fin_Color.rgb = fin_Color.rgb * (1.0 - lay_Color.a) + lay_Color.rgb;
```
Comment (jeapostrophe) [inline mode-lambda/backend/gl/combine.fragment.glsl]: Can you reformat this so only the one line changes, as it is it looks like nothing is different.

File: mode-lambda/backend/gl/ngl.fragment.glsl [hunk 2]
```diff
@@ -21,12 +21,13 @@ void main(void)
...
+    texture(SpriteAtlasTex,
+            vec2(TexCoord.x / texSize.x,
+                 TexCoord.y / texSize.y))
     // But it doesn't work on some ES devices I have, so we do this instead:
-    texelFetch(SpriteAtlasTex,
-               ivec2( trunc(TexCoord.x), trunc(TexCoord.y) ),
-               0)
+    //texelFetch(SpriteAtlasTex,
+    //           ivec2( trunc(TexCoord.x), trunc(TexCoord.y) ),
+    //           0)
```
Comment (jeapostrophe) [inline mode-lambda/backend/gl/ngl.fragment.glsl]: Could you put these changes with an `if true/false` and then I can go in and make it compile-time?

File: mode-lambda/backend/gl/ngl.fragment.glsl [hunk 3]
```diff
@@ -43,6 +44,7 @@ void main(void)
 
   fin_Color.a = PixelColor.a * Color.a;
   fin_Color.rgb = PixelColor.rgb + Color.rgb;
+  fin_Color.rgb = min(fin_Color.rgb, fin_Color.a);
```
Comment (jeapostrophe) [inline mode-lambda/backend/gl/ngl.fragment.glsl]: What is the logic here?

File: mode-lambda/backend/gl/ngl.vertex.glsl [hunk 4]
```diff
@@ -26,11 +26,11 @@ void main(void)
...
   ivec2 spec_v = vertexSpec[spec_idx];
-  int xcoeff = spec_v.x;
-  int ycoeff = spec_v.y;
+  int xcoeff = spec_v.x;  // used for wrapping
+  int ycoeff = spec_v.y;  // used for wrapping
   vec2 hv = compute_center_iTexCoord();
-  float horiz = hv.x;
-  float vert = hv.y;
+  float horiz = hv.x; // -1 to +1
+  float vert = hv.y;  // -1 to +1
```
Comment (jeapostrophe) [inline mode-lambda/backend/gl/ngl.vertex.glsl]: Could you remove these comments, since it is just diff noise?

File: mode-lambda/backend/gl/ngl.vertex.glsl [hunk 5]
```diff
@@ -72,6 +72,14 @@ void main(void)
...
   }
 
+  // make sure that we don't end up just between pixels
+  // otherwise you can't see a 1x1 sprite given integer coords
+  // because almostPosn goes from -0.5 to 0.5 which means no pixel
+  // hits it, so no fragment shader is run (I think)
+  //almostPosn.x = trunc(almostPosn.x);
+  //almostPosn.y = trunc(almostPosn.y);
+
+  // move back to normalized (-1 to +1) coords for output
```
Comment (jeapostrophe) [inline mode-lambda/backend/gl/ngl.vertex.glsl]: Same here

File: mode-lambda/backend/gl/util.rkt [hunk 6]
```diff
@@ -65,7 +65,9 @@
   (vector GL_TEXTURE0 GL_TEXTURE1 GL_TEXTURE2
           GL_TEXTURE3 GL_TEXTURE4 GL_TEXTURE5
           GL_TEXTURE6 GL_TEXTURE7 GL_TEXTURE8
-          GL_TEXTURE9 GL_TEXTURE10 GL_TEXTURE11))
+          GL_TEXTURE9 GL_TEXTURE10 GL_TEXTURE11
+          GL_TEXTURE12 GL_TEXTURE13 GL_TEXTURE14
+          GL_TEXTURE15))
```
Comment (jeapostrophe) [inline mode-lambda/backend/gl/util.rkt]: I pushed a change to this based on reading the spec about the invariants on these constants

File: mode-lambda/static.rkt [hunk 7]
```diff
@@ -141,6 +143,22 @@
...
+    ; For debugging, force the alpha to be at least 1 so that
+    ; when we write-png-bytes!, set-argb-pixels will pay attention
+    ; to the rgb values (it ignores them if alpha is 0
+    ;(bytes-set! atlas-bs idx (max 1 a))
+
+    (for [(k (in-range 1 4))]
+      (define col (bytes-ref atlas-bs (+ idx k)))
+      (bytes-set! atlas-bs (+ idx k)
+                  (fl->fx (flround (fl/ (fx->fl (* a col)) 255.0))))))
+
```
Comment (jeapostrophe) [inline mode-lambda/static.rkt:160]: Why do we ever NOT want premultiplied alpha?

File: mode-lambda/static.rkt [hunk 8]
```diff
@@ -149,7 +167,7 @@
                  racket/draw)
   (define root-bm (make-bitmap w h))
   (send root-bm set-argb-pixels 0 0 w h bs)
-  (send root-bm save-file p 'png 100 #:unscaled? #t))
+  (send root-bm save-file p 'png))
```
Comment (jeapostrophe) [inline mode-lambda/static.rkt]: Was this causing problems?

**Comment by jeapostrophe** [comment]
I hope you still feel like a good sport, because I just messed you up again. I changed the shader to use a texture array rather than an array of textures and tested with 256 layers. You'll need to modify your side again.

File: mode-lambda/backend/gl/ngl.fragment.glsl (hunk 3 above)
Comment (david-vanderson) [reply mode-lambda/backend/gl/ngl.fragment.glsl]: The intent is to make fin_Color a proper premultiplied alpha color after adding Color.rgb.  Otherwise I see artifacts when combining layers.  I'm not 100% sure though - it depends on what you intend to happen when adding #:r/g/b to a sprite.  Does that make sense?

File: mode-lambda/static.rkt (hunk 7 above)
Comment (david-vanderson) [reply mode-lambda/static.rkt:160]: I think we always want it.

File: mode-lambda/static.rkt (hunk 8 above)
Comment (david-vanderson) [reply mode-lambda/static.rkt]: no, sorry

**Comment by david-vanderson** [comment]
no problem, will fix issues and resubmit

**Comment by david-vanderson** [comment]
Thanks for the layers change!  Hope these new commits address the problems.

I'm unsure what needs to be done for palette colors with premultiplied alpha.  Probably 'compile-sprite-db' should convert them too, but I don't use them so I'm worried about doing it wrong.

Thanks!

File: mode-lambda/backend/gl/combine.fragment.glsl [hunk 9]
```diff
@@ -59,7 +59,7 @@ void main() {
         && (0.0 <= ex && ex <= width)) {
       fin_Color.rgb =
           fin_Color.rgb * (1.0 - lay_Color.a)
-        + lay_Color.rgb * lay_Color.a;
+        + lay_Color.rgb;
```
Comment (jeapostrophe) [inline mode-lambda/backend/gl/combine.fragment.glsl]: If I am reading the last equation here --- https://en.wikipedia.org/wiki/Alpha_compositing#Description --- correctly then we can do `fin_Color = lay_Color + fin_Color * (1 - lay_Color.a)` and make sure that the content of the layer is also premultiplied-alpha. (The current code doesn't change .a)

File: mode-lambda/backend/gl/ngl.fragment.glsl (hunk 3 above)
Comment (jeapostrophe) [reply mode-lambda/backend/gl/ngl.fragment.glsl]: I think I understand now.

We want #:rgb to apply an opaque mask, which the add operation does on the rgb values. However, we also want the #:a to change the transparency. In the language of the original paper on premultipled alpha --- https://keithp.com/~keithp/porterduff/p253-porter.pdf ---

```
Let r, g, b, a be the parameters
PixelColor = what the sprite atlas says (premultiplied alpha)
ColorMask = (r, g, b, 1) // A fully opaque color mask
FinalColor = dissolve( plus( PixelColor, ColorMask ), a)
```

So converting that into GLSL:

```
dissolveFactor = Color.a;
vec4 opaqueMask = (Color.rgb, 1.0);
fin_Color = dissolveFactor * (PixelColor + opaqueMask);
```

Can you tell me if that looks right to you?

File: mode-lambda/static.rkt [hunk 10]
```diff
@@ -141,6 +143,22 @@
     (hash-set! spr->idx spr pi)
     (vector-set! idx->w*h*tx*ty pi (vector w h tx ty)))
 
+
+  ; convert to premultiplied alpha
+  (for [(i (in-range (/ (bytes-length atlas-bs) 4)))]
+    (define idx (* 4 i))
```
Comment (jeapostrophe) [inline mode-lambda/static.rkt]: `([idx (in-range 0 (bytes-length atlas-bs) 4)])`

File: mode-lambda/static.rkt (hunk 7 above)
Comment (jeapostrophe) [reply mode-lambda/static.rkt:160]: Yes, I agree now.

File: mode-lambda/backend/gl/ngl.fragment.glsl [hunk 11]
```diff
@@ -19,15 +19,19 @@ void main(void)
   vec4 SpriteColor;
 
   vec2 texSize = textureSize(SpriteAtlasTex, 0);
-  SpriteColor =
+  if (true) {
```
Comment (jeapostrophe) [inline mode-lambda/backend/gl/ngl.fragment.glsl]: I think that this `true` path is the same as what this code used to be, but I only changed it because an ES device was weird. Maybe this `true` should be `@(if (gl-es?) "false" "true")`, because this change shouldn't have anything to do with the smoothing.

File: mode-lambda/backend/gl.rkt [hunk 12]
```diff
@@ -87,7 +87,10 @@
       ;; xxx allow these to be updated
       (define SpriteAtlasId (make-2dtexture))
       (with-texture (GL_TEXTURE0 SpriteAtlasId)
-        (load-texture/bytes atlas-size atlas-size atlas-bs))
+        (load-texture/bytes atlas-size atlas-size atlas-bs)
+        (glGenerateMipmap GL_TEXTURE_2D)
```
Comment (jeapostrophe) [inline mode-lambda/backend/gl.rkt]: I am happy making this optional, but if you want to, then add a parameter like `gl-filter-mode` (Maybe `gl-smoothing?`) and add it to the call to `define-make-delayed-render` next to `gl-filter-mode` then add an extra argument to `make-draw` after `screen-mode` and just add a `(when smooth? ....)` around these three lines.

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: mode-lambda/backend/gl/combine.fragment.glsl (hunk 9 above)
Comment (david-vanderson) [reply mode-lambda/backend/gl/combine.fragment.glsl]: I think you are right - I'll change it and test.

File: mode-lambda/backend/gl.rkt (hunk 12 above)
Comment (david-vanderson) [reply mode-lambda/backend/gl.rkt]: I will attempt to do that.

File: mode-lambda/backend/gl/ngl.fragment.glsl (hunk 11 above)
Comment (david-vanderson) [reply mode-lambda/backend/gl/ngl.fragment.glsl]: I thought calling texture() did smoothing (if enabled) while calling texelfetch() did not?

File: mode-lambda/backend/gl/ngl.fragment.glsl (hunk 3 above)
Comment (david-vanderson) [reply mode-lambda/backend/gl/ngl.fragment.glsl]: I'm less sure now.  Using it as an opaque mask destroys the alpha value of the pixel, right?  That would be a big change.  Generally I'm using #:rgb to colorize black text and lines that have transparency, but I'm less sure how it should work on non-black pixels.
Say #:rgb is (1, 0, 0) and #:a is 0.9
Example 1: black sprite pixel is (0, 0, 0, 0.3) (rgba)
-> I think we would want the final color to be (0.3, 0, 0, 0.3) * 0.9
Example 2: white sprite pixel is (0.3, 0.3, 0.3, 0.3)
-> Should the final color be an average of red and white? ((0.3 + 0.3) / 2, 0.3 / 2, 0.3 / 2) * 0.9

File: mode-lambda/backend/gl/ngl.fragment.glsl (hunk 11 above)
Comment (jeapostrophe) [reply mode-lambda/backend/gl/ngl.fragment.glsl]: That's right, but then this means that we can just enable/disable smoothing in that one spot and not affect the shader.

Comment (david-vanderson) [reply mode-lambda/backend/gl/ngl.fragment.glsl]: I'm confused.  The shader has to be affected to know whether to call texture() or texelFetch().  Or maybe you are saying we always call texture() unless it is a gl-es? thing, in which case we give up on smoothing?  That makes sense to me - I will do that in the next attempt.

File: mode-lambda/static.rkt (hunk 10 above)
Comment (david-vanderson) [reply mode-lambda/static.rkt]: doh - of course.  Thanks!

File: mode-lambda/backend/gl/ngl.fragment.glsl (hunk 11 above)
Comment (jeapostrophe) [reply mode-lambda/backend/gl/ngl.fragment.glsl]: Yes, that's what I mean. (Maybe this ES device I tested on was just busted anyways.)

**Comment by david-vanderson** [comment]
I noticed during testing that generating the mipmaps of the sprite atlas can bleed sprites into each other.  Adding more padding around each sprite helps, and it's the only idea I have.  Do you have a better one?

File: mode-lambda/backend/gl/ngl.fragment.glsl (hunk 3 above)
Comment (david-vanderson) [reply mode-lambda/backend/gl/ngl.fragment.glsl]: I think what I have now is essentially what was there originally but converted to premultiplied alpha.  Originally adding the rgb values together could go over 255 but I think glsl automatically caps it.  So that's where capping the output rgb to the output alpha came from.  Still not totally sure though.

**Comment by jeapostrophe** [comment]
The bleeding is almost certainly coming from reading the top corner of the texel rather than the middle, so when you interpolate and read the adjacent texels, they are merged together. You basically want to add 0.5 to the arguments to the texture call. See if that helps. (This isn't necessary with the texelFetch, because you're reading a particular texel there.)

I found this picture of what's happening: https://gamedev.stackexchange.com/a/49585

**Comment by david-vanderson** [comment]
Good idea, but as best I can tell the texture calls are working correctly.  For example, if I turn smoothing off, and then add 0.5 to the texture call, I see only a half pixel on the top and left of a sprite.  I think I'm seeing the same thing with smoothing on.  Also as I add more and more padding around each sprite, I see the problem only at smaller and smaller zooms.

But I did remember you saying at Racketcon that we could do the mipmapping manually, and I think I now understand what you meant.  Since we know the zoom of the sprite, we could dynamically select a pre-rendered smaller version, and generate power-of-2 smaller sizes (down to 2x2) during sprite atlas compilation.  Does that sound like a viable plan to you?  If so, should the dynamic sprite selection happen in the vertex shader or can we do it somewhere in racket?

**Comment by jeapostrophe** [comment]
Selecting the sprite based on the zoom isn't going to more IMHO because you can zoom the layer separately from the sprite, so you need to know the two of those.

I was thinking of having separate mipmapped sprite atlases by changing the way the way that (for example) pict and 2htdp/image objects are turned into pixels and rendering those at a few levels. Then all the sprite ids would be same and you'd actually run the original rendering algorithm for the sprite at higher resolution.

**Comment by david-vanderson** [comment]
Could we select the specific sprite atlas based on the sprite+layer zoom in the vertex shader?

Also I thought mipmapping was entirely about *lower* resolutions than the given sprites.  Are we talking about different use cases?

**Comment by jeapostrophe** [comment]
I could imagine storing all the many sprite atlases as a texture array and selecting them on some criteria, but that's just what mipmaps do, so better let OpenGL be in charge.

Lower resolution is isomorphic to bigger resolutions, because you are just changing your perspective on what the "real" level is.

**Comment by david-vanderson** [comment]
That makes sense to me.  I've been doing some more testing, and I'm only seeing the mipmapping issue visually in some very corner cases. So I think this pull request is feature complete.  Do you have any more comments on implementation?

**Comment by jeapostrophe** [comment]
Alright, if you think it's good, I'll merge it. Thanks for the hard work!

**Comment by david-vanderson** [comment]
Thank you so much!
